interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,966 @@
+			delete(variantMap, variantStr)
+			// Record a notice for the offending rule, just in case
+			rdb.notifyRule(user, variantEntry.RuleID, nil)
+			continue
+		}
+		expired, err := matchingRule.Expired(currTime)
+		switch {
+		case err != nil:
+			// Issue with expiration, this should not occur
+			logger.Noticef("error while checking whether rule had expired: %v", err)
+			fallthrough
+		case expired:
+			continue
+		}
+		// Need to compare the path pattern variant, not the rule's path
+		// pattern, so that only variants which match are included,
+		// and the highest precedence variant can be computed.
+		matched, err := patterns.PathPatternMatches(variantStr, path)
+		if err != nil {
+			// Only possible error is ErrBadPattern, which should not occur
+			return false, err
+		}
+		if matched {
+			matchingVariants = append(matchingVariants, variantEntry.Variant)
+		}
+	}
+	if len(matchingVariants) == 0 {
+		return false, ErrNoMatchingRule
+	}
+	highestPrecedenceVariant, err := patterns.HighestPrecedencePattern(matchingVariants, path)
+	if err != nil {
+		return false, err
+	}
+	matchingEntry := variantMap[highestPrecedenceVariant.String()]
+	matchingID := matchingEntry.RuleID
+	matchingRule, err := rdb.ruleWithID(matchingID)
+	if err != nil {
+		// Database was left inconsistent, should not occur
+		return false, ErrRuleIDNotFound
+	}
+	return matchingRule.Outcome.AsBool()
+}
+
+// ruleWithIDForUser returns the rule with the given ID, if it exists, for the
+// given user. Otherwise, returns an error.
+//
+// The caller must ensure that the database lock is held.
+func (rdb *RuleDB) ruleWithIDForUser(user uint32, id prompting.IDType) (*Rule, error) {
+	rule, err := rdb.ruleWithID(id)
+	if err != nil {
+		return nil, err
+	}
+	if rule.User != user {
+		return nil, ErrUserNotAllowed
+	}
+	return rule, nil
+}
+
+// RuleWithID returns the rule with the given ID.
+// If the rule is not found, returns ErrRuleNotFound.
+// If the rule does not apply to the given user, returns ErrUserNotAllowed.
+func (rdb *RuleDB) RuleWithID(user uint32, id prompting.IDType) (*Rule, error) {
+	rdb.mutex.RLock()
+	defer rdb.mutex.RUnlock()
+	return rdb.ruleWithIDForUser(user, id)
+}
+
+// Creates a rule with the given information and adds it to the rule database.
+// If any of the given parameters are invalid, returns an error. Otherwise,
+// returns the newly-added rule, and saves the database to disk.
+func (rdb *RuleDB) AddRule(user uint32, snap string, iface string, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (*Rule, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	newRule, err := rdb.populateNewRule(user, snap, iface, constraints, outcome, lifespan, duration)
+	if err != nil {
+		return nil, err
+	}
+	if needToSave, err, conflicts, conflictingPermission := rdb.addRuleToTree(newRule); err != nil {
+		if needToSave {
+			rdb.save() // discard any new error, which should not occur
```

Comment:
why would we try to save if we got an error? this is a very odd API

Reply:
The rationale was that if in the process of trying (and eventually failing) to add a rule, there was an existing rule which was expired and conflicted with the new rule, we needed to remove that expired rule immediately. Expired rules are pruned on load. My worry is that there are a few messy scenarios:
- (change (notice) -> save -> load (no notice)): one notice, this is what's desired
- (change (notice) -> no save -> load (notice)): duplicate notices for the expired rule
- (change (no notice) -> save -> load (no notice)): no notice, very bad!!
- (change (no notice) -> no save -> load (notice)): one notice, this is what's desired in the unusual case

So we need to record the notice immediately, else there's risk of losing the notice. And I think there's not much harm in duplicate notices. So let's just not save when operations fail. In most cases, we'll save later before loading, but in the rare case where we don't worst case we get duplicate notices for the expired rule, which is really fine.